File: tests/integration/tap_stack_integration_test.go
package integration_test

import (
	"context"
	"encoding/json"
	"os"
	"testing"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/cloudformation"
	"github.com/aws/aws-sdk-go-v2/service/codebuild"
	"github.com/aws/aws-sdk-go-v2/service/codepipeline"
	"github.com/aws/aws-sdk-go-v2/service/iam"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/aws/aws-sdk-go-v2/service/sns"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// Outputs represents the stack outputs from deployment
type Outputs struct {
	PipelineArn          string `json:"PipelineArn"`
	BuildProjectName     string `json:"BuildProjectName"`
	ArtifactBucketName   string `json:"ArtifactBucketName"`
	NotificationTopicArn string `json:"NotificationTopicArn"`
}

// loadOutputs loads the cfn-outputs/flat-outputs.json file
func loadOutputs(t *testing.T) Outputs {
	data, err := os.ReadFile("../../cfn-outputs/flat-outputs.json")
	require.NoError(t, err, "Failed to read ../../cfn-outputs/flat-outputs.json")

	var outputs Outputs
	err = json.Unmarshal(data, &outputs)
	require.NoError(t, err, "Failed to parse outputs JSON")

	return outputs
}

func TestIntegration_S3Bucket(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	outputs := loadOutputs(t)
	require.NotEmpty(t, outputs.ArtifactBucketName, "ArtifactBucketName not found in outputs")

	// Setup AWS config
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
	require.NoError(t, err, "Failed to load AWS config")

	// Create S3 client
	s3Client := s3.NewFromConfig(cfg)

	// Test: Verify bucket exists
	t.Run("bucket exists", func(t *testing.T) {
		_, err := s3Client.HeadBucket(context.TODO(), &s3.HeadBucketInput{
			Bucket: &outputs.ArtifactBucketName,
		})
		assert.NoError(t, err, "Bucket should exist")
	})

	// Test: Verify bucket versioning
	t.Run("bucket has versioning enabled", func(t *testing.T) {
		versioning, err := s3Client.GetBucketVersioning(context.TODO(), &s3.GetBucketVersioningInput{
			Bucket: &outputs.ArtifactBucketName,
		})
		require.NoError(t, err)
		assert.Equal(t, "Enabled", string(versioning.Status), "Versioning should be enabled")
	})

	// Test: Verify bucket encryption
	t.Run("bucket has encryption enabled", func(t *testing.T) {
		encryption, err := s3Client.GetBucketEncryption(context.TODO(), &s3.GetBucketEncryptionInput{
			Bucket: &outputs.ArtifactBucketName,
		})
		require.NoError(t, err)
		assert.NotEmpty(t, encryption.ServerSideEncryptionConfiguration.Rules, "Encryption rules should exist")
	})

	// Test: Verify public access block
	t.Run("bucket blocks public access", func(t *testing.T) {
		publicAccess, err := s3Client.GetPublicAccessBlock(context.TODO(), &s3.GetPublicAccessBlockInput{
			Bucket: &outputs.ArtifactBucketName,
		})
		require.NoError(t, err)
		assert.True(t, *publicAccess.PublicAccessBlockConfiguration.BlockPublicAcls, "Should block public ACLs")
		assert.True(t, *publicAccess.PublicAccessBlockConfiguration.BlockPublicPolicy, "Should block public policy")
		assert.True(t, *publicAccess.PublicAccessBlockConfiguration.IgnorePublicAcls, "Should ignore public ACLs")
		assert.True(t, *publicAccess.PublicAccessBlockConfiguration.RestrictPublicBuckets, "Should restrict public buckets")
	})
}

func TestIntegration_SNSTopic(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	outputs := loadOutputs(t)
	require.NotEmpty(t, outputs.NotificationTopicArn, "NotificationTopicArn not found in outputs")

	// Setup AWS config
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
	require.NoError(t, err, "Failed to load AWS config")

	// Create SNS client
	snsClient := sns.NewFromConfig(cfg)

	// Test: Verify topic exists
	t.Run("topic exists", func(t *testing.T) {
		attrs, err := snsClient.GetTopicAttributes(context.TODO(), &sns.GetTopicAttributesInput{
			TopicArn: &outputs.NotificationTopicArn,
		})
		require.NoError(t, err, "Topic should exist")
		assert.NotEmpty(t, attrs.Attributes, "Topic attributes should not be empty")
	})
}

func TestIntegration_CodeBuildProject(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	outputs := loadOutputs(t)
	require.NotEmpty(t, outputs.BuildProjectName, "BuildProjectName not found in outputs")

	// Setup AWS config
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
	require.NoError(t, err, "Failed to load AWS config")

	// Create CodeBuild client
	cbClient := codebuild.NewFromConfig(cfg)

	// Test: Verify project exists
	t.Run("project exists", func(t *testing.T) {
		projects, err := cbClient.BatchGetProjects(context.TODO(), &codebuild.BatchGetProjectsInput{
			Names: []string{outputs.BuildProjectName},
		})
		require.NoError(t, err, "Failed to get project")
		require.Len(t, projects.Projects, 1, "Should find exactly one project")
		assert.Equal(t, outputs.BuildProjectName, *projects.Projects[0].Name)
	})

	// Test: Verify project configuration
	t.Run("project has correct configuration", func(t *testing.T) {
		projects, err := cbClient.BatchGetProjects(context.TODO(), &codebuild.BatchGetProjectsInput{
			Names: []string{outputs.BuildProjectName},
		})
		require.NoError(t, err)

		project := projects.Projects[0]
		assert.Equal(t, "BUILD_GENERAL1_SMALL", string(project.Environment.ComputeType))
		assert.Equal(t, "LINUX_CONTAINER", string(project.Environment.Type))
		assert.False(t, *project.Environment.PrivilegedMode)
	})
}

func TestIntegration_CodePipeline(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	outputs := loadOutputs(t)
	require.NotEmpty(t, outputs.PipelineArn, "PipelineArn not found in outputs")

	// Setup AWS config
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
	require.NoError(t, err, "Failed to load AWS config")

	// Create CodePipeline client
	cpClient := codepipeline.NewFromConfig(cfg)

	// Extract pipeline name from ARN
	// ARN format: arn:aws:codepipeline:region:account:pipeline-name
	// We need to extract the last part after the last colon
	pipelineName := ""
	for i := len(outputs.PipelineArn) - 1; i >= 0; i-- {
		if outputs.PipelineArn[i] == ':' {
			pipelineName = outputs.PipelineArn[i+1:]
			break
		}
	}
	require.NotEmpty(t, pipelineName, "Failed to extract pipeline name from ARN")

	// Test: Verify pipeline exists
	t.Run("pipeline exists", func(t *testing.T) {
		pipeline, err := cpClient.GetPipeline(context.TODO(), &codepipeline.GetPipelineInput{
			Name: &pipelineName,
		})
		require.NoError(t, err, "Pipeline should exist")
		assert.NotNil(t, pipeline.Pipeline)
	})

	// Test: Verify pipeline has three stages
	t.Run("pipeline has three stages", func(t *testing.T) {
		pipeline, err := cpClient.GetPipeline(context.TODO(), &codepipeline.GetPipelineInput{
			Name: &pipelineName,
		})
		require.NoError(t, err)

		assert.Len(t, pipeline.Pipeline.Stages, 3, "Pipeline should have three stages")
		assert.Equal(t, "Source", *pipeline.Pipeline.Stages[0].Name)
		assert.Equal(t, "Build", *pipeline.Pipeline.Stages[1].Name)
		assert.Equal(t, "Deploy", *pipeline.Pipeline.Stages[2].Name)
	})

	// Test: Verify pipeline status
	t.Run("pipeline is in valid state", func(t *testing.T) {
		state, err := cpClient.GetPipelineState(context.TODO(), &codepipeline.GetPipelineStateInput{
			Name: &pipelineName,
		})
		require.NoError(t, err)
		assert.NotEmpty(t, state.StageStates, "Pipeline should have stage states")
	})
}

func TestIntegration_IAMRoles(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	outputs := loadOutputs(t)

	// Setup AWS config
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
	require.NoError(t, err, "Failed to load AWS config")

	// Get environment suffix from outputs (extract from bucket name)
	envSuffix := ""
	bucketName := outputs.ArtifactBucketName
	if len(bucketName) > len("cicd-artifacts-") {
		envSuffix = bucketName[len("cicd-artifacts-"):]
	}
	require.NotEmpty(t, envSuffix, "Failed to extract environment suffix")

	// Create IAM client
	iamClient := iam.NewFromConfig(cfg)

	// Test: Verify CodeBuild role exists
	t.Run("CodeBuild role exists", func(t *testing.T) {
		roleName := "codebuild-role-" + envSuffix
		role, err := iamClient.GetRole(context.TODO(), &iam.GetRoleInput{
			RoleName: &roleName,
		})
		require.NoError(t, err, "CodeBuild role should exist")
		assert.Contains(t, *role.Role.AssumeRolePolicyDocument, "codebuild.amazonaws.com")
	})

	// Test: Verify CodePipeline role exists
	t.Run("CodePipeline role exists", func(t *testing.T) {
		roleName := "codepipeline-role-" + envSuffix
		role, err := iamClient.GetRole(context.TODO(), &iam.GetRoleInput{
			RoleName: &roleName,
		})
		require.NoError(t, err, "CodePipeline role should exist")
		assert.Contains(t, *role.Role.AssumeRolePolicyDocument, "codepipeline.amazonaws.com")
	})
}

func TestIntegration_StackOutputs(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	outputs := loadOutputs(t)

	// Test: All required outputs exist
	t.Run("all required outputs exist", func(t *testing.T) {
		assert.NotEmpty(t, outputs.PipelineArn, "PipelineArn should exist")
		assert.NotEmpty(t, outputs.BuildProjectName, "BuildProjectName should exist")
		assert.NotEmpty(t, outputs.ArtifactBucketName, "ArtifactBucketName should exist")
		assert.NotEmpty(t, outputs.NotificationTopicArn, "NotificationTopicArn should exist")
	})

	// Test: Output format validation
	t.Run("outputs have correct format", func(t *testing.T) {
		assert.Contains(t, outputs.PipelineArn, "arn:aws:codepipeline:")
		assert.Contains(t, outputs.NotificationTopicArn, "arn:aws:sns:")
		assert.Contains(t, outputs.ArtifactBucketName, "cicd-artifacts-")
		assert.Contains(t, outputs.BuildProjectName, "build-project-")
	})
}

func TestIntegration_ResourceConnectivity(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping integration test in short mode")
	}

	// Setup AWS config
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-east-1"))
	require.NoError(t, err, "Failed to load AWS config")

	// Test: Pipeline can access artifact bucket
	t.Run("pipeline has access to artifact bucket", func(t *testing.T) {
		cfnClient := cloudformation.NewFromConfig(cfg)

		// Get stack name from environment
		envSuffix := os.Getenv("ENVIRONMENT_SUFFIX")
		if envSuffix == "" {
			envSuffix = "dev"
		}
		stackName := "TapStack" + envSuffix

		// Get stack resources
		resources, err := cfnClient.DescribeStackResources(context.TODO(), &cloudformation.DescribeStackResourcesInput{
			StackName: &stackName,
		})

		if err == nil {
			// Verify both pipeline and bucket exist in the stack
			var hasPipeline, hasBucket bool
			for _, resource := range resources.StackResources {
				if *resource.ResourceType == "AWS::CodePipeline::Pipeline" {
					hasPipeline = true
				}
				if *resource.ResourceType == "AWS::S3::Bucket" {
					hasBucket = true
				}
			}
			assert.True(t, hasPipeline, "Stack should contain CodePipeline")
			assert.True(t, hasBucket, "Stack should contain S3 bucket")
		}
	})
}

File: test/tap_stack_integration_test.py
"""
Integration tests for CI/CD Pipeline Infrastructure
Tests validate actual deployed AWS resources and their configurations
"""

import json
import os
import boto3
import pytest
from botocore.exceptions import ClientError

# Get deployment outputs
OUTPUTS_FILE = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'cfn-outputs', 'flat-outputs.json')

@pytest.fixture(scope='module')
def deployment_outputs():
    """Load deployment outputs from flat-outputs.json"""
    if not os.path.exists(OUTPUTS_FILE):
        pytest.skip("Deployment outputs not found")

    with open(OUTPUTS_FILE, 'r') as f:
        return json.load(f)

@pytest.fixture(scope='module')
def aws_region():
    """Get AWS region from environment or default"""
    return os.environ.get('AWS_REGION', 'us-east-1')

@pytest.fixture(scope='module')
def s3_client(aws_region):
    """Create S3 client"""
    return boto3.client('s3', region_name=aws_region)

@pytest.fixture(scope='module')
def dynamodb_client(aws_region):
    """Create DynamoDB client"""
    return boto3.client('dynamodb', region_name=aws_region)

@pytest.fixture(scope='module')
def codepipeline_client(aws_region):
    """Create CodePipeline client"""
    return boto3.client('codepipeline', region_name=aws_region)

@pytest.fixture(scope='module')
def codebuild_client(aws_region):
    """Create CodeBuild client"""
    return boto3.client('codebuild', region_name=aws_region)

@pytest.fixture(scope='module')
def sns_client(aws_region):
    """Create SNS client"""
    return boto3.client('sns', region_name=aws_region)

@pytest.fixture(scope='module')
def cloudwatch_logs_client(aws_region):
    """Create CloudWatch Logs client"""
    return boto3.client('logs', region_name=aws_region)

@pytest.fixture(scope='module')
def cloudwatch_events_client(aws_region):
    """Create CloudWatch Events client"""
    return boto3.client('events', region_name=aws_region)

@pytest.fixture(scope='module')
def codestar_client(aws_region):
    """Create CodeStar Connections client"""
    return boto3.client('codestar-connections', region_name=aws_region)

@pytest.fixture(scope='module')
def iam_client(aws_region):
    """Create IAM client"""
    return boto3.client('iam', region_name=aws_region)


class TestS3Resources:
    """Test S3 bucket resources"""

    def test_artifact_bucket_exists(self, s3_client, deployment_outputs):
        """Test that pipeline artifacts bucket exists"""
        bucket_name = deployment_outputs.get('artifact_bucket_name')
        assert bucket_name is not None, "Artifact bucket name not in outputs"

        try:
            response = s3_client.head_bucket(Bucket=bucket_name)
            assert response['ResponseMetadata']['HTTPStatusCode'] == 200
        except ClientError as e:
            pytest.fail(f"Artifact bucket {bucket_name} does not exist: {e}")

    def test_state_bucket_exists(self, s3_client, deployment_outputs):
        """Test that Terraform state bucket exists"""
        bucket_name = deployment_outputs.get('state_bucket_name')
        assert bucket_name is not None, "State bucket name not in outputs"

        try:
            response = s3_client.head_bucket(Bucket=bucket_name)
            assert response['ResponseMetadata']['HTTPStatusCode'] == 200
        except ClientError as e:
            pytest.fail(f"State bucket {bucket_name} does not exist: {e}")

    def test_artifact_bucket_encryption(self, s3_client, deployment_outputs):
        """Test that artifact bucket has encryption enabled"""
        bucket_name = deployment_outputs['artifact_bucket_name']

        try:
            response = s3_client.get_bucket_encryption(Bucket=bucket_name)
            rules = response['ServerSideEncryptionConfiguration']['Rules']
            assert len(rules) > 0, "No encryption rules configured"
            assert rules[0]['ApplyServerSideEncryptionByDefault']['SSEAlgorithm'] in ['AES256', 'aws:kms']
        except ClientError as e:
            pytest.fail(f"Failed to get encryption config for {bucket_name}: {e}")

    def test_state_bucket_encryption(self, s3_client, deployment_outputs):
        """Test that state bucket has encryption enabled"""
        bucket_name = deployment_outputs['state_bucket_name']

        try:
            response = s3_client.get_bucket_encryption(Bucket=bucket_name)
            rules = response['ServerSideEncryptionConfiguration']['Rules']
            assert len(rules) > 0, "No encryption rules configured"
            assert rules[0]['ApplyServerSideEncryptionByDefault']['SSEAlgorithm'] in ['AES256', 'aws:kms']
        except ClientError as e:
            pytest.fail(f"Failed to get encryption config for {bucket_name}: {e}")

    def test_artifact_bucket_versioning(self, s3_client, deployment_outputs):
        """Test that artifact bucket has versioning enabled"""
        bucket_name = deployment_outputs['artifact_bucket_name']

        try:
            response = s3_client.get_bucket_versioning(Bucket=bucket_name)
            assert response.get('Status') == 'Enabled', "Versioning not enabled"
        except ClientError as e:
            pytest.fail(f"Failed to get versioning config for {bucket_name}: {e}")

    def test_state_bucket_versioning(self, s3_client, deployment_outputs):
        """Test that state bucket has versioning enabled"""
        bucket_name = deployment_outputs['state_bucket_name']

        try:
            response = s3_client.get_bucket_versioning(Bucket=bucket_name)
            assert response.get('Status') == 'Enabled', "Versioning not enabled"
        except ClientError as e:
            pytest.fail(f"Failed to get versioning config for {bucket_name}: {e}")

    def test_buckets_block_public_access(self, s3_client, deployment_outputs):
        """Test that buckets block public access"""
        buckets = [
            deployment_outputs['artifact_bucket_name'],
            deployment_outputs['state_bucket_name']
        ]

        for bucket_name in buckets:
            try:
                response = s3_client.get_public_access_block(Bucket=bucket_name)
                config = response['PublicAccessBlockConfiguration']
                assert config['BlockPublicAcls'] == True
                assert config['BlockPublicPolicy'] == True
                assert config['IgnorePublicAcls'] == True
                assert config['RestrictPublicBuckets'] == True
            except ClientError as e:
                pytest.fail(f"Failed to get public access block for {bucket_name}: {e}")


class TestDynamoDBResources:
    """Test DynamoDB resources"""

    def test_state_lock_table_exists(self, dynamodb_client, deployment_outputs):
        """Test that state lock table exists"""
        table_name = deployment_outputs.get('state_lock_table_name')
        assert table_name is not None, "State lock table name not in outputs"

        try:
            response = dynamodb_client.describe_table(TableName=table_name)
            assert response['Table']['TableStatus'] in ['ACTIVE', 'UPDATING']
        except ClientError as e:
            pytest.fail(f"State lock table {table_name} does not exist: {e}")

    def test_state_lock_table_billing(self, dynamodb_client, deployment_outputs):
        """Test that state lock table uses PAY_PER_REQUEST billing"""
        table_name = deployment_outputs['state_lock_table_name']

        try:
            response = dynamodb_client.describe_table(TableName=table_name)
            billing_mode = response['Table']['BillingModeSummary']['BillingMode']
            assert billing_mode == 'PAY_PER_REQUEST', f"Expected PAY_PER_REQUEST, got {billing_mode}"
        except ClientError as e:
            pytest.fail(f"Failed to get billing mode for {table_name}: {e}")

    def test_state_lock_table_key_schema(self, dynamodb_client, deployment_outputs):
        """Test that state lock table has correct key schema"""
        table_name = deployment_outputs['state_lock_table_name']

        try:
            response = dynamodb_client.describe_table(TableName=table_name)
            key_schema = response['Table']['KeySchema']
            assert len(key_schema) == 1, "Expected 1 key"
            assert key_schema[0]['AttributeName'] == 'LockID'
            assert key_schema[0]['KeyType'] == 'HASH'
        except ClientError as e:
            pytest.fail(f"Failed to get key schema for {table_name}: {e}")

    def test_state_lock_table_pitr(self, dynamodb_client, deployment_outputs):
        """Test that point-in-time recovery is enabled"""
        table_name = deployment_outputs['state_lock_table_name']

        try:
            response = dynamodb_client.describe_continuous_backups(TableName=table_name)
            pitr_status = response['ContinuousBackupsDescription']['PointInTimeRecoveryDescription']['PointInTimeRecoveryStatus']
            assert pitr_status == 'ENABLED', f"PITR not enabled, status: {pitr_status}"
        except ClientError as e:
            pytest.fail(f"Failed to get PITR status for {table_name}: {e}")


class TestCodePipelineResources:
    """Test CodePipeline resources"""

    def test_pipeline_exists(self, codepipeline_client, deployment_outputs):
        """Test that pipeline exists"""
        pipeline_name = deployment_outputs.get('pipeline_name')
        assert pipeline_name is not None, "Pipeline name not in outputs"

        try:
            response = codepipeline_client.get_pipeline(name=pipeline_name)
            assert response['pipeline']['name'] == pipeline_name
        except ClientError as e:
            pytest.fail(f"Pipeline {pipeline_name} does not exist: {e}")

    def test_pipeline_stages(self, codepipeline_client, deployment_outputs):
        """Test that pipeline has all required stages"""
        pipeline_name = deployment_outputs['pipeline_name']

        try:
            response = codepipeline_client.get_pipeline(name=pipeline_name)
            stages = response['pipeline']['stages']
            stage_names = [stage['name'] for stage in stages]

            required_stages = ['Source', 'Validate', 'Plan', 'Approval', 'Apply']
            for required_stage in required_stages:
                assert required_stage in stage_names, f"Stage {required_stage} not found"
        except ClientError as e:
            pytest.fail(f"Failed to get pipeline stages: {e}")

    def test_pipeline_manual_approval_stage(self, codepipeline_client, deployment_outputs):
        """Test that pipeline has manual approval stage configured"""
        pipeline_name = deployment_outputs['pipeline_name']

        try:
            response = codepipeline_client.get_pipeline(name=pipeline_name)
            stages = response['pipeline']['stages']

            approval_stage = [s for s in stages if s['name'] == 'Approval']
            assert len(approval_stage) == 1, "Approval stage not found"

            actions = approval_stage[0]['actions']
            assert len(actions) > 0, "No actions in Approval stage"
            assert actions[0]['actionTypeId']['category'] == 'Approval'
            assert actions[0]['actionTypeId']['provider'] == 'Manual'
        except ClientError as e:
            pytest.fail(f"Failed to verify approval stage: {e}")


class TestCodeBuildResources:
    """Test CodeBuild resources"""

    def test_validate_project_exists(self, codebuild_client, deployment_outputs):
        """Test that validate CodeBuild project exists"""
        project_name = deployment_outputs.get('validate_project_name')
        assert project_name is not None, "Validate project name not in outputs"

        try:
            response = codebuild_client.batch_get_projects(names=[project_name])
            assert len(response['projects']) == 1
            assert response['projects'][0]['name'] == project_name
        except ClientError as e:
            pytest.fail(f"Validate project {project_name} does not exist: {e}")

    def test_plan_project_exists(self, codebuild_client, deployment_outputs):
        """Test that plan CodeBuild project exists"""
        project_name = deployment_outputs.get('plan_project_name')
        assert project_name is not None, "Plan project name not in outputs"

        try:
            response = codebuild_client.batch_get_projects(names=[project_name])
            assert len(response['projects']) == 1
            assert response['projects'][0]['name'] == project_name
        except ClientError as e:
            pytest.fail(f"Plan project {project_name} does not exist: {e}")

    def test_apply_project_exists(self, codebuild_client, deployment_outputs):
        """Test that apply CodeBuild project exists"""
        project_name = deployment_outputs.get('apply_project_name')
        assert project_name is not None, "Apply project name not in outputs"

        try:
            response = codebuild_client.batch_get_projects(names=[project_name])
            assert len(response['projects']) == 1
            assert response['projects'][0]['name'] == project_name
        except ClientError as e:
            pytest.fail(f"Apply project {project_name} does not exist: {e}")

    def test_codebuild_projects_have_logs(self, codebuild_client, deployment_outputs):
        """Test that CodeBuild projects have CloudWatch Logs configured"""
        project_names = [
            deployment_outputs['validate_project_name'],
            deployment_outputs['plan_project_name'],
            deployment_outputs['apply_project_name']
        ]

        try:
            response = codebuild_client.batch_get_projects(names=project_names)
            for project in response['projects']:
                logs_config = project.get('logsConfig', {})
                cloudwatch_logs = logs_config.get('cloudWatchLogs', {})
                assert cloudwatch_logs.get('status') == 'ENABLED', f"CloudWatch Logs not enabled for {project['name']}"
        except ClientError as e:
            pytest.fail(f"Failed to verify CodeBuild logs configuration: {e}")


class TestSNSResources:
    """Test SNS resources"""

    def test_sns_topic_exists(self, sns_client, deployment_outputs):
        """Test that SNS topic exists"""
        topic_arn = deployment_outputs.get('sns_topic_arn')
        assert topic_arn is not None, "SNS topic ARN not in outputs"

        try:
            response = sns_client.get_topic_attributes(TopicArn=topic_arn)
            assert response['Attributes']['TopicArn'] == topic_arn
        except ClientError as e:
            pytest.fail(f"SNS topic {topic_arn} does not exist: {e}")

    def test_sns_topic_has_policy(self, sns_client, deployment_outputs):
        """Test that SNS topic has access policy"""
        topic_arn = deployment_outputs['sns_topic_arn']

        try:
            response = sns_client.get_topic_attributes(TopicArn=topic_arn)
            policy = response['Attributes'].get('Policy')
            assert policy is not None, "SNS topic policy not set"

            policy_doc = json.loads(policy)
            assert 'Statement' in policy_doc, "Policy has no statements"
            assert len(policy_doc['Statement']) > 0, "Policy has no statements"
        except ClientError as e:
            pytest.fail(f"Failed to get SNS topic policy: {e}")


class TestCodeStarResources:
    """Test CodeStar Connections resources"""

    def test_codestar_connection_exists(self, codestar_client, deployment_outputs):
        """Test that CodeStar Connection exists"""
        connection_arn = deployment_outputs.get('codestar_connection_arn')
        assert connection_arn is not None, "CodeStar connection ARN not in outputs"

        try:
            response = codestar_client.get_connection(ConnectionArn=connection_arn)
            assert response['Connection']['ConnectionArn'] == connection_arn
        except ClientError as e:
            pytest.fail(f"CodeStar connection {connection_arn} does not exist: {e}")

    def test_codestar_connection_provider(self, codestar_client, deployment_outputs):
        """Test that CodeStar Connection is for GitHub"""
        connection_arn = deployment_outputs['codestar_connection_arn']

        try:
            response = codestar_client.get_connection(ConnectionArn=connection_arn)
            provider_type = response['Connection']['ProviderType']
            assert provider_type == 'GitHub', f"Expected GitHub provider, got {provider_type}"
        except ClientError as e:
            pytest.fail(f"Failed to get CodeStar connection provider: {e}")


class TestCloudWatchResources:
    """Test CloudWatch resources"""

    def test_validate_log_group_exists(self, cloudwatch_logs_client):
        """Test that validate log group exists"""
        log_group_pattern = '/aws/codebuild/validate-'

        try:
            response = cloudwatch_logs_client.describe_log_groups(logGroupNamePrefix=log_group_pattern)
            assert len(response['logGroups']) > 0, f"No log groups found with prefix {log_group_pattern}"
        except ClientError as e:
            pytest.fail(f"Failed to find validate log group: {e}")

    def test_plan_log_group_exists(self, cloudwatch_logs_client):
        """Test that plan log group exists"""
        log_group_pattern = '/aws/codebuild/plan-'

        try:
            response = cloudwatch_logs_client.describe_log_groups(logGroupNamePrefix=log_group_pattern)
            assert len(response['logGroups']) > 0, f"No log groups found with prefix {log_group_pattern}"
        except ClientError as e:
            pytest.fail(f"Failed to find plan log group: {e}")

    def test_apply_log_group_exists(self, cloudwatch_logs_client):
        """Test that apply log group exists"""
        log_group_pattern = '/aws/codebuild/apply-'

        try:
            response = cloudwatch_logs_client.describe_log_groups(logGroupNamePrefix=log_group_pattern)
            assert len(response['logGroups']) > 0, f"No log groups found with prefix {log_group_pattern}"
        except ClientError as e:
            pytest.fail(f"Failed to find apply log group: {e}")

    def test_pipeline_failure_event_rule_exists(self, cloudwatch_events_client):
        """Test that pipeline failure event rule exists"""
        rule_pattern = 'pipeline-failure-'

        try:
            response = cloudwatch_events_client.list_rules(NamePrefix=rule_pattern)
            assert len(response['Rules']) > 0, f"No event rules found with prefix {rule_pattern}"
        except ClientError as e:
            pytest.fail(f"Failed to find pipeline failure event rule: {e}")


class TestResourceNaming:
    """Test resource naming conventions"""

    def test_resource_names_include_suffix(self, deployment_outputs):
        """Test that all resource names include environment suffix"""
        # Extract suffix from pipeline name
        pipeline_name = deployment_outputs['pipeline_name']
        # Pipeline name format: terraform-pipeline-{suffix}
        suffix = pipeline_name.split('-')[-1]

        assert suffix, "Could not extract environment suffix from pipeline name"

        # Check all resource names include the suffix
        resource_names = [
            deployment_outputs['pipeline_name'],
            deployment_outputs['artifact_bucket_name'],
            deployment_outputs['state_bucket_name'],
            deployment_outputs['state_lock_table_name'],
            deployment_outputs['validate_project_name'],
            deployment_outputs['plan_project_name'],
            deployment_outputs['apply_project_name']
        ]

        for resource_name in resource_names:
            assert suffix in resource_name, f"Resource {resource_name} does not include suffix {suffix}"
